### PR TITLE
feat(ir): more flexible dereferencing for join right-hand side

### PR DESCRIPTION
**Pull request by kszucs**
The join dereferencing logic required explicit references to the exact right hand side table which disallowed more convenient syntaxes like `t1.join(t2.at_time(t1.c), [t1.a == t2.d])`. 

Resolves https://github.com/ibis-project/ibis/issues/8581

### CODE REVIEW

**Comment by kszucs** [comment]
There is a more complicated test case in the impala test suite, inspecting it.

**Comment by kszucs** [comment]
The multi-table dereference mapping calculation must be slightly smarter by tracking the distance from the target relations and picking the closest one (unless there are multiple ones at the same distance in which case an ambiguity error should be raised).

**Comment by cpcloud** [comment]
@kszucs The two failures are pretty easy to fix. Otherwise, is this PR ready to go in?

**Comment by gforsyth** [comment]
CI is green if @kszucs is happy with this.

File: ibis/backends/impala/tests/snapshots/test_sql/test_join_aliasing/out.sql [hunk 1]
```diff
@@ -62,7 +62,7 @@ INNER JOIN (
       GROUP BY
         1
     ) AS `t10`
-      ON `t8`.`d` = `t10`.`d`
+      ON `t10`.`d` = `t10`.`d`
```
Comment (cpcloud) [inline ibis/backends/impala/tests/snapshots/test_sql/test_join_aliasing/out.sql:65]: This doesn't look correct to me. Should be `t8` as it was before.

**Review by cpcloud — CHANGES_REQUESTED** [review]

**Comment by kszucs** [comment]
No, I am not happy with this yet.

**Comment by cpcloud** [comment]
Should we close this in favor of #8992?

**Comment by kszucs** [comment]
Yes

**Pull request closed without merging**